src/main/java/org/springframework/data/hazelcast/repository/query/HazelcastPartTreeQuery.java

```diff
@@ -200,34 +200,27 @@ private Object executeFindQuery(final KeyValueQuery<?> query, final QueryMethod
 	private Object executePageSliceQuery(final Object[] parameters, final KeyValueQuery<?> query,
 			final QueryMethod queryMethod) {
 		long totalElements = -1;
-
 		int indexOfPageRequest = queryMethod.getParameters().getPageableIndex();
 		Pageable pageRequest = (Pageable) parameters[indexOfPageRequest];
 
-		/* TODO Eliminate count call for Slice, retrieve "rows+1" instead to determine if next page exists.
-		 */
-		if (query.getCritieria() == null) {
-			totalElements = this.keyValueOperations.count(queryMethod.getEntityInformation().getJavaType());
-		} else {
-			totalElements = this.keyValueOperations.count(query, queryMethod.getEntityInformation().getJavaType());
-		}
-
 		int requiredRows = pageRequest.getPageSize();
 
 		query.setOffset(pageRequest.getOffset());
-		query.setRows(pageRequest.getPageSize());
+		query.setRows(queryMethod.isSliceQuery() ? requiredRows + 1 : requiredRows);
 
 		Iterable<?> resultSet = this.keyValueOperations.find(query, queryMethod.getEntityInformation().getJavaType());
 		List<?> content = IterableConverter.toList(resultSet);
 
 		if (queryMethod.isPageQuery()) {
+			if (query.getCritieria() == null) {
+				totalElements = this.keyValueOperations.count(queryMethod.getEntityInformation().getJavaType());
+			} else {
+				totalElements = this.keyValueOperations.count(query, queryMethod.getEntityInformation().getJavaType());
+			}
 			return new PageImpl(content, pageRequest, totalElements);
 		} else {
-			boolean hasNext = totalElements > (query.getOffset() + query.getRows());
```

Comment:
I would explicitly add `if queryMethod.isSliceQuery()` for more readability or We can put a small comment indicating that this part is sliceQuery.. We can even put `query.setRows` logic into their respective if block...

Reply:
I can add a comment since this method only called for page or slice queries.